File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int _speed; 
    private int _battery = 100;
    private int _batteryDrain;
    private int _distanceDriven = 0;
    
    public RemoteControlCar(int speed, int batteryDrain) 
    {
        _speed = speed; 
        _batteryDrain = batteryDrain;
    }
    
    public bool BatteryDrained()
        => _battery < _batteryDrain;

    public int DistanceDriven()
        => _distanceDriven;

    public void Drive()
    {
        if (_battery >= _batteryDrain)
        {
            _battery -= _batteryDrain;
            _distanceDriven += _speed;
        }
    }

    public static RemoteControlCar Nitro()
        => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int _distance;

    public RaceTrack(int distance) => _distance = distance;

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained())
        {
            car.Drive();
        }
        return car.DistanceDriven() >= _distance;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int speed;
    private int batteryCharge = 100;
    private int batteryDrain;
    private int distanceDriven = 0;

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryCharge < batteryDrain;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (batteryCharge >= batteryDrain) {
            batteryCharge -= batteryDrain;
            distanceDriven += speed;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private final int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while (!car.batteryDrained())
            car.drive();
        return car.distanceDriven() >= distance;
    }
}
